plugins/inputs/xtremio/xtremio.go

```diff
@@ -0,0 +1,388 @@
+		}
+	}
+	
+	tlsCfg, err := xio.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = tlsCfg
+
+	return nil
+}
+
+func (xio *XtremIO) Gather(acc telegraf.Accumulator) error {
+	var err = xio.authenticate()
+	if err != nil {
+		return err
+	}
+	if xio.Cookie == nil {
+		return fmt.Errorf("no authentication cookie set")
+	}
+
+	var wg sync.WaitGroup
+	for _, collector := range xio.Collectors {
+		wg.Add(1)
+		go func(collector string) {
+			defer wg.Done()
+
+			resp, err := xio.call(collector)
+			if err != nil {
+				acc.AddError(err)
+				return
+			}
+
+			// Due to an inconsistency in the XtremIO API, the XMS endpoint
+			// returns a json array with XMSS as the result. Which is why this
+			// if statement here exists
+			var arr []gjson.Result
+			if collector == "xms" {
+				arr = gjson.Get(resp, "xmss").Array()
+			}else{
+				arr = gjson.Get(resp, collector).Array()
+			}
+			
+			for _, item := range arr {
+				itemSplit := strings.Split(gjson.Get(item.Raw, "href").Str, "/")
+				url := ""
+				if len(itemSplit) < 1 {
+					continue
+				}
+				url = collector + "/" + itemSplit[len(itemSplit)-1]
+
+				switch collector {
+				case "bbus":
+					wg.Add(1)
+					go xio.gatherBBUs(acc, url, &wg)
+				case "clusters":
+					wg.Add(1)
+					go xio.gatherClusters(acc, url, &wg)
+				case "ssds":
+					wg.Add(1)
+					go xio.gatherSSDs(acc, url, &wg)
+				case "volumes":
+					wg.Add(1)
+					go xio.gatherVolumes(acc, url, &wg)
+				case "xms":
+					wg.Add(1)
+					go xio.gatherXMS(acc, url, &wg)
+				default:
+					acc.AddError(fmt.Errorf("Specified Collector Isn't Supported: " + collector))
+				}
+			}
+		}(collector)
+	}
+	wg.Wait()
+
+	xio.Cookie = nil
+
+	return nil
+}
+
+func (xio *XtremIO) gatherBBUs(acc telegraf.Accumulator, url string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	resp, err := xio.call(url)
+	if err != nil {
+		acc.AddError(err)
+	}
+	tags := map[string]string{
+		"serial_number": gjson.Get(resp, "content.serial-number").Str,
+		"guid":          gjson.Get(resp, "content.guid").Str,
+		"power_feed":    gjson.Get(resp, "content.power-feed").Str,
+		"name":          gjson.Get(resp, "content.name").Str,
+		"model_name":    gjson.Get(resp, "content.model-name").Str,
+	}
+	fields := make(map[string]interface{})
+	fields["bbus_power"], err = strconv.Atoi(gjson.Get(resp, "content.power").Raw)
+	if err != nil {acc.AddError(err)}
+
+	fields["bbus_average_daily_temp"], err = strconv.Atoi(gjson.Get(resp, "content.avg-daily-temp").Raw)
+	if err != nil {acc.AddError(err)}
+
+	fields["bbus_enabled"] = (gjson.Get(resp, "content.enabled-state").Str == "enabled")
+	fields["bbus_ups_need_battery_replacement"] = (gjson.Get(resp, "content.ups-need-battery-replacement").Str == "true")
+	fields["bbus_ups_low_battery_no_input"] = (gjson.Get(resp, "content.is-low-battery-no-input").Str == "true")
+
+	acc.AddFields("xio", fields, tags)
+}
```

Comment:
This whole approach is a bit unconventional. Usually, we define a `struct` with the expected fields and add the JSON tags mapping to the JSON structure. Then, just `json.Unmarshal` the received text into the struct and access the fields.

Reply:
I'll work on refactoring it to use structs. I did it this way mostly because I don't like how Go handles JSON. I've been spoiled by Python and Powershell.